### PR TITLE
dkg: add signature verification by server

### PR DESCRIPTION
**Pull request by dB2510**
Adds cluster-definition hash signature verification by server sent by client.

category: feature
ticket: #684

### CODE REVIEW

**Comment by OisinKyne** [comment]
What server is this? The relayer or the leader or something else?

**Comment by dB2510** [comment]
> What server is this? The relayer or the leader or something else?

@OisinKyne So we are introducing server and client terms in dkg sync service. More on this [here](https://docs.google.com/document/d/1nQwOWfGui6dzAbET3VLjHIc-3_sNzohTDpEu4d0nRyU/edit#heading=h.mssb0822zapb).

Basically, right now every peer opens streams to receive pings from other peers and open new streams to send pings. In this sync flow implementation, we are dividing receiving part as Server and sending part as Client. Where client(other peers) will send their definition hash signature signed using their `charon-enr-private-key` to server and server will verify to ensure that everyone is aligned with the same cluster-definition before starting DKG ceremony.

File: dkg/sync/client.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ func (*Client) Shutdown() error {
 }
 
 // NewClient starts a goroutine that establishes a long lived connection to a p2p server and returns a new Client instance.
-func NewClient(ctx context.Context, tcpNode host.Host, server p2p.Peer, hash []byte, onFailure func(), ch chan *pb.MsgSyncResponse) Client {
+func NewClient(ctx context.Context, tcpNode host.Host, server p2p.Peer, hashSig []byte, onFailure func(), ch chan *pb.MsgSyncResponse) Client {
```
Comment (xenowits) [inline dkg/sync/client.go:54]: discussion: what does `hashSig` mean? signature on hash, `sigHash`?

File: dkg/sync/server.go [hunk 2]
```diff
@@ -90,12 +90,29 @@ func NewServer(ctx context.Context, tcpNode host.Host, peers []p2p.Peer, hash []
 			return
 		}
 
+		pID := s.Conn().RemotePeer()
+		log.Debug(ctx, "Message received from client", z.Any("peer", p2p.PeerName(pID)))
+
+		peerPubkey, err := pID.ExtractPublicKey()
```
Comment (xenowits) [inline dkg/sync/server.go:96]: nit: can just name it `pubKey`

File: dkg/sync/server.go [hunk 3]
```diff
@@ -90,12 +90,29 @@ func NewServer(ctx context.Context, tcpNode host.Host, peers []p2p.Peer, hash []
...
+		if err != nil {
+			log.Error(ctx, "Get client's public key", err)
```
Comment (xenowits) [inline dkg/sync/server.go:98]: nit:
```suggestion
			log.Error(ctx, "Get client public key", err)
```

File: dkg/sync/server.go [hunk 4]
```diff
@@ -54,7 +54,7 @@ func (*Server) AwaitAllShutdown() error {
 }
 
 // NewServer registers a Stream Handler and returns a new Server instance.
-func NewServer(ctx context.Context, tcpNode host.Host, peers []p2p.Peer, hash []byte, onFailure func()) *Server {
+func NewServer(ctx context.Context, tcpNode host.Host, peers []p2p.Peer, definitionHash []byte, onFailure func()) *Server {
```
Comment (xenowits) [inline dkg/sync/server.go:57]:
```suggestion
func NewServer(ctx context.Context, tcpNode host.Host, peers []p2p.Peer, defHash []byte, onFailure func()) *Server {
```

**Review by xenowits — APPROVED** [review]

File: dkg/sync/client.go (hunk 1 above)
Comment (dB2510) [reply dkg/sync/client.go:54]: hashSig is a signature of config_hash signed using charon-enr-private-key

File: dkg/sync/server.go (hunk 2 above)
Comment (dB2510) [reply dkg/sync/server.go:96]: sure

**Pull request merged**